normalize mixed/absent entities with a configurable “snakenull” label

jordandekraker opened the pull request:
## Proposed changes

Adds optional normalization so that when a component lists an entity in wildcards but some matched files lack that entity, missing values are relabeled to `snakenull` (configurable). If an entity listed in wildcards is entirely absent across the dataset, that entity is skipped for that component. Path construction can include or omit the {key}-snakenull segment via a flag. Default remains the current behavior (feature disabled). This integrates cleanly with `generate_inputs()`/`BidsComponent` (per Snakebids docs), and preserves expand/wildcards semantics users rely on. 

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you are unsure about any of the choices, don't hesitate to ask!

- [x] Changes have been tested to ensure that fix is effective or that a feature works.
- [x] Changes pass the unit tests (pytest -q snakebids/tests/test_snakenull.py;  poe quality; still working through other quality checks)
- [x] I have included necessary documentation or comments (as necessary)
- [x] Any dependent changes have been merged and published

## Notes
All PRs will undergo the unit testing before being reviewed. You may be requested to explain or make additional changes before the PR is accepted.

jordandekraker committed:
relatively complete initial commit thanks to chatgpt

jordandekraker committed:
lint

jordandekraker committed:
lint (even in tests! so thorough)

jordandekraker committed:
allow input_generation to proceed even with missing entities

jordandekraker committed:
call normalize_inputs_with_snakenull within generate_inputs (so the user doesn't have to in their own snakefile)

jordandekraker committed:
cleanup of functions and unused stuff

jordandekraker committed:
lint

jordandekraker committed:
loosened pyright; 0 errors now

jordandekraker committed:
added proper pulling of config info; will lint later

jordandekraker committed:
lint

akhanf commented:
Yes some upstream env issue is not causing the tests to run, can you try running the tests locally? 

The changes are hard to review since the diff is quite messy -- are you able to provide some high-level summary of the implementation-level changes?

jordandekraker commented:
Maybe you already looked at it now, but here's some further explanation fo the feature from my locally built RTD:
<img width="823" height="840" alt="image" src="https://github.com/user-attachments/assets/8b69f683-a0c4-48ee-9fc9-1a9e90af665d" />

This is specifically in response to https://github.com/khanlab/snakebids/issues/469 

Note that I have basically no experience with Typing or pyright. Maybe someone who knows this well and has time can handle this, but otherwise I'll learn. But before I do I just wanted some general feedback. You seemed skeptical in the https://github.com/khanlab/snakebids/issues/469 discussion so i just wanted to make a quick sketchup for you to consider. 

I can try to clean up the messy diffs too, this was mostly an artifact of having chatgpt work on the initial draft.

akhanf commented:
> Maybe you already looked at it now, but here's some further explanation fo the feature from my locally built RTD:
> 
> This is specifically in response to #469

I read through that and understand what you mean there, it is more a sketch of the implementation that was missing for me. I also don't see how this doesn't break snakebids, since how do you grab bids dataset inputs with expansion BidsComponent.path, now that you don't just have a single unique path pattern? 

I haven't tried running this PR but I imagine the actual tests will be failing, unless you've tried them already? 

> Note that I have basically no experience with Typing or pyright. Maybe someone who knows this well and has time can handle this, but otherwise I'll learn. But before I do I just wanted some general feedback. You seemed skeptical in the #469 discussion so i just wanted to make a quick sketchup for you to consider.
> 
> I can try to clean up the messy diffs too, this was mostly an artifact of having chatgpt work on the initial draft.


Overall I really just want to make sure the code quality in snakebids is maintained --  alot of effort has been put into this in the past,  and proposals for major overhauls like this are typically sketched out in discussion and conceptualized and debated first before just diving in..

jordandekraker commented:
Ah OK I understand - I'm not used to working with such a high set of development standards :grimacing:

My dryrun testing within micapipe works, generating a clean DAG. 

The logic/sketchup of this feature follows something like this:
1. Discover & parse files
generate_inputs queries PyBIDS and, per matched file, parses only the requested wildcards that exist on that file.
We collect:
  - a set of candidate path templates (with {wildcards}), and
  - a per-file dict of parsed entity values (missing → "").
2. (Conditional) collapse template shapes
If snakenull.enabled is true (globally or per component), we collapse filename shapes by removing optional tag-{tag} tokens that aren’t present for all files (e.g., drop acq-{acq} when only some files have it). If disabled, we keep legacy behavior and raise the classic “Multiple path templates...” ConfigError.
3. Build rectangular zip_lists aligned to the final template
We only include placeholders that actually appear in the chosen template and pad each file’s row so all lists have equal length.
4. Normalize entities (snakenull pass)
If snakenull.enabled is true, we compute entity domains from the per-file dicts:
- Drop entities that are never present across the component (they’re ignored downstream).
- For entities present in some files but missing in others, insert the placeholder label (default snakenull) into that entity’s value set.
- We do not inject snakenull per-file; we add it to the domain so downstream iteration can include it as a value.
- For frozen objects (e.g., BidsComponent), we avoid mutating attrs and stash read-only context instead.

Chatgpt also had these optimistic comments on what we've done here:
Edge cases & guarantees:
- Works with custom_path components (normalizer falls back to zip_lists inspection).
- Collapse only affects filename tokens (keeps directory entities like ses-{session} intact).
- zip_lists are always rectangular and consistent with the final path template.
- Backwards compatible when disabled; tests added for enabled/disabled/mixed cases.

Why this is safe
- We never change which files are included—just how entity domains are exposed when data is heterogeneous.
- When disabled, behavior is identical to prior releases.

zihuaihuai committed:
pyright snakenull

zihuaihuai committed:
Apply ruff formatting fixes

jordandekraker committed:
Merge pull request #471 from zihuaihuai/fix-failing-tests

Fix failing tests

zihuaihuai committed:
fix: resolve entity/tag mapping regression in input generation

- Fix incorrect entity-to-wildcard mapping in _safe_parse_per_file
- Fix zip_lists construction in _build_zip_lists_from_parsed
- Ensure consistent mapping between entity names and wildcard names
- Resolves test_entity_excluded_when_filter_false failure

Fixes regression introduced by snakenull integration where entity names
and wildcard names were incorrectly mapped, causing empty strings in
zip_lists instead of proper entity values.

jordandekraker committed:
Merge pull request #472 from zihuaihuai/fix-failing-tests

Fix incorrect entity-to-wildcard mapping in _safe_parse_per_file
Fix zip_lists construction in _build_zip_lists_from_parsed
Ensure consistent mapping between entity names and wildcard names
Resolves test_entity_excluded_when_filter_false failure
